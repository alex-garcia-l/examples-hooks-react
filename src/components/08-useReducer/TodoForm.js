import React, { useRef } from 'react';
import Swal from 'sweetalert2';
import PropTypes from 'prop-types';

import { useForm } from '../../hooks/useForm';

const swalWithBootstrapButtons = Swal.mixin({
  customClass: {
    confirmButton: 'btn btn-success',
    cancelButton: 'btn btn-outline-danger ms-4'
  },
  buttonsStyling: false
})

export const TodoForm = ({ handleAdd }) => {

  const inputDescription = useRef();
  const [{ description }, handleInputChange, resetForm] = useForm({
    description: ''
  });

  const handleSubmit = (evt) => {
    evt.preventDefault();

    if (description === '') {
      swalWithBootstrapButtons.fire({
        title: 'Something is wrong!',
        text: 'Add a description',
        icon: 'error',
        confirmButtonText: 'Accept',
        timer: 2000,
        timerProgressBar: true,
        didOpen: (toast) => {
          toast.addEventListener('mouseenter', Swal.stopTimer)
          toast.addEventListener('mouseleave', Swal.resumeTimer)
        }
      });
      inputDescription.current.focus();
      return;
    }

    handleAdd({
      id: new Date().getTime(),
      desciption: description,
      done: false
    });

    resetForm();
    inputDescription.current.focus();
  }

  return (
    <>
      <h3>Add Todo</h3>
      <form onSubmit={handleSubmit}>
        <div className="mb-3">
          <label htmlFor="description" className="form-label">Description</label>
          <input
            type="text"
            autoFocus
            className="form-control"
            placeholder="Description"
            autoComplete="off"
            name="description"
            value={description}
            onChange={handleInputChange}
            ref={inputDescription}
          />
          {/* <div id="emailHelp" className="form-text">We'll never share your email with anyone else.</div> */}
        </div>

        <div className="d-grid gap-2">
          <button
            type="submit"
            className="btn btn-primary d-block"
          >
            Add
          </button>
        </div>
      </form>
    </>
  );
};

TodoForm.propTypes = {
  handleAdd: PropTypes.func.isRequired
}
