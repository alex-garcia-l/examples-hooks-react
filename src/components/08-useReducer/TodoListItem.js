import React from 'react';
import PropTypes from 'prop-types';
import { Trash2, Check, X } from 'react-feather';

export const TodoListItem = ({ todo, index, handleToggle, handleDelete }) => {
  return (
    <tr className={`${todo.done ? 'text-dark bg-light text-decoration-line-through fst-italic' : ''}`}>
      <th>{index + 1}</th>
      <th>{todo.desciption}</th>
      <th>
        <button
          className={`btn ${!todo.done ? 'btn-success' : 'btn-warning'} btn-sm me-2`}
          onClick={() => handleToggle(todo.id)}
        >
          {!todo.done ? <Check size={16} /> : <X size={16} />}
        </button>
        <button
          className="btn btn-danger btn-sm"
          onClick={() => handleDelete(todo)}
        >
          <Trash2 size={16} />
        </button>
      </th>
    </tr>
  );
};

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired
}
