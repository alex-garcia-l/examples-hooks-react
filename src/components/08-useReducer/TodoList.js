import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleToggle, handleDelete }) => {
  return (
    <>
      <h3>Todo List ({todos.length} totals)</h3>
      <table className="table">
        <thead>
          <tr>
            <th>#</th>
            <th>Description</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {
            todos
              .sort(function (a, b) { return a.done - b.done })
              .map((todo, index) => (
                <TodoListItem
                  key={todo.id}
                  todo={todo}
                  index={index}
                  handleToggle={handleToggle}
                  handleDelete={handleDelete}
                />
              ))
          }
        </tbody>
      </table>
    </>
  );
};
