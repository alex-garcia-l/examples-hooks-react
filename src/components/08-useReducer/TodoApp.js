import React, { useEffect, useReducer } from 'react';
import Swal from 'sweetalert2';

import { TodoForm } from './TodoForm';
import { TodoList } from './TodoList';
import { useReducerTodo } from './useReducerTodo';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

const swalWithBootstrapButtons = Swal.mixin({
  customClass: {
    confirmButton: 'btn btn-success',
    cancelButton: 'btn btn-outline-danger ms-4'
  },
  buttonsStyling: false
});

const Toast = Swal.mixin({
  toast: true,
  position: 'top-end',
  showConfirmButton: false,
  timer: 3000,
  timerProgressBar: true,
  didOpen: (toast) => {
    toast.addEventListener('mouseenter', Swal.stopTimer)
    toast.addEventListener('mouseleave', Swal.resumeTimer)
  }
});

export const TodoApp = () => {

  const [todos, dispatchTodos] = useReducer(useReducerTodo, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleAdd = (newTodo) => {
    dispatchTodos({
      type: 'add',
      payload: newTodo
    });
    
    Toast.fire({
      icon: 'success',
      title: 'Successfully added'
    });
  }

  const handleToggle = (id) => {
    dispatchTodos({
      type: 'toggle-done',
      payload: id
    });
  }

  const handleDelete = (todo) => {
    swalWithBootstrapButtons.fire({
      title: `Are you sure to remove <span class="fw-bolder">${todo.desciption}</span>?`,
      text: "You won't be able to revert this!",
      icon: 'warning',
      showCancelButton: true,
      confirmButtonText: 'Yes, delete it!',
      cancelButtonText: 'No, cancel!',
    }).then((result) => {
      if (result.isConfirmed) {
        dispatchTodos({
          type: 'delete',
          payload: todo.id
        });
      }
    });
  }

  return (
    <div className="">
      <div className="row">
        <h1>Todo App</h1>
        <hr />
      </div>
      <div className="row">
        <div className='col-12 col-md-5 col-lg-4'>
          <TodoForm handleAdd={handleAdd} />
        </div>
        <div className='col-12 col-md-7 col-lg-8 mt-4 mt-md-0'>
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          />
        </div>
      </div>
    </div>
  );
};
