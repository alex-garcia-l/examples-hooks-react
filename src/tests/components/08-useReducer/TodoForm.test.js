import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import { TodoForm } from '../../../components/08-useReducer/TodoForm';

describe('Pruebas para el componente <TodoApp />', () => {

  const handleAdd = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <TodoForm
        handleAdd={handleAdd}
      />
    );
  });

  test('debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('NO debe de llamarse la función handleAdd', () => {

    const formSubmit = wrapper.find('form').prop('onSubmit');

    act(() => {
      formSubmit({ preventDefault() { } });
    })

    expect(handleAdd).toHaveBeenCalledTimes(0);

  });

  test('debe de llamarse la función handleAdd', () => {

    const value = 'Aprender react';
    wrapper.find('input').simulate('change', {
      target: {
        name: 'description',
        value: value
      }
    });

    const formSubmit = wrapper.find('form').prop('onSubmit');    
    
    act(() => {
      formSubmit({ preventDefault() { } });
    });

    expect(handleAdd).toHaveBeenCalledTimes(1);
    expect(handleAdd).toHaveBeenCalledWith({
      id: expect.any(Number),
      desciption: value,
      done: false
    });

    // expect(wrapper.find('input').prop('value')).toBe(''); // TODO: terminar de hacer
  });

});
