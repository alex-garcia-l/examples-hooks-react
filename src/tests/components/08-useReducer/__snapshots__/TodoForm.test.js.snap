// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para el componente <TodoApp /> debe mostrarse correctamente 1`] = `
Array [
  <h3>
    Add Todo
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="description"
      >
        Description
      </label>
      <input
        autoComplete="off"
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-primary d-block"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>,
]
`;
